### PR TITLE
[MPS] Add native strided API for MPSNDArray starting with macOS 15

### PR DESCRIPTION
**Pull request by DenisVieriu97**
Add support for native strides in MPS starting with macOS Sequoia. This will get rid of the additional gather and scatter operations needed to solve the strides or storage offsets of the tensors. 

Summary of changes (starting with macOS 15):
- Add support for **MPS strided API** (strides/storage offsets etc):
   - [initWithBuffer:offset:descriptor:](https://developer.apple.com/documentation/metalperformanceshaders/mpsndarray/4391636-initwithbuffer?language=objc)
   - [arrayViewWithCommandBuffer:descriptor:aliasing:](https://developer.apple.com/documentation/metalperformanceshaders/mpsndarray/3114040-arrayviewwithcommandbuffer?language=objc)
   - [arrayViewWithShape:strides:](https://developer.apple.com/documentation/metalperformanceshaders/mpsndarray/4408694-arrayviewwithshape?language=objc)
   - [reshapeWithCommandBuffer:sourceArray:shape:destinationArray:](https://developer.apple.com/documentation/metalperformanceshaders/mpsndarrayidentity/4438557-reshapewithcommandbuffer?language=objc)
- Add native support for NHWC convolutions (without incurring any extra copy from NCHW -> NHWC -> NCHW).
- Add support for strided output buffers (previously we would create a contiguous buffer  

OSes older than macOS 15 will run the old gather/scatter code path to solve strides/storage offsets. 

--- 

Couple performance stats collected from torchbench comparing macOS 15 vs macOS 14: 
```
- test_train[functorch_maml_omniglot-mps]: 27% faster
- test_train[timm_vision_transformer-mps]: 12% faster
- test_train[hf_T5-mps]: 9.46% faster
```

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: test/test_mps.py [hunk 1]
```diff
@@ -3543,6 +3547,116 @@ def test_slice(self):
...
+        r_mps = z_mps + 1
+        self.assertEqual(r, r_mps)
+
+    def test_slice_reshape_view_api_test_2(self):
+        x = torch.randn(2, 4)
+        x_mps = x.detach().clone().to("mps")
+        ones = torch.ones(4)
+        ones_mps = torch.ones(4, device="mps")
+
+        print(x_mps)
```
Comment (albanD) [inline test/test_mps.py:3573]: print?

File: test/test_mps.py [hunk 2]
```diff
@@ -3543,6 +3547,116 @@ def test_slice(self):
         mps_slice4 = mps_x[1, :].to('cpu')
         self.assertEqual(cpu_slice4, mps_slice4)
 
+    def test_slice_reshape_view_api_test_1(self):
```
Comment (albanD) [inline test/test_mps.py:3550]: These tests seems to have a lot of things in common.
1) are they not covered by OpInfo based testing already?
2) If not, should they be refactored such that they have a lot less boiler plate and consistent output metadata checking?
3) While making them generic, you can use parametrize decorator to make them run on all dtypes (avoiding int-specific tests)

File: aten/src/ATen/native/mps/operations/Indexing.mm [hunk 3]
```diff
@@ -608,17 +608,20 @@ Tensor index_select_mps(const Tensor& self, int64_t dim, const Tensor& index) {
       newCachedGraph->outputTensor_ = outputTensor;
     });
 
+    // MPS TODO: MPS Gather is failing with MPS strided API. Fallback to old gather.
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/Indexing.mm:611]: There are a couple TODOs lying around, should we open issues to track them down?

File: aten/src/ATen/native/mps/operations/Copy.mm [hunk 4]
```diff
@@ -89,12 +89,12 @@ static void copy_cast_mps(at::Tensor& dst,
   Tensor dst = dst_;
   Tensor src = src_;
 
-  if (!dst_.is_contiguous(MemoryFormat::Contiguous) && !sameMemFormat) {
+  if (dst_.strides() != src_.strides()) {
```
Comment (albanD) [inline aten/src/ATen/native/mps/operations/Copy.mm:92]: That looks suspicious? 
Anything that has stride but not storage_offset feels suspicious to me :D

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 5]
```diff
@@ -359,26 +370,152 @@ void printTensorNDArray(const Tensor& t) {
...
+    NSUInteger axis1 = srcRank - i - 1;
+    NSUInteger axis2 = dimensionOrder.end() - axisIter - 1;
+    iter_swap(dimensionOrder.begin() + i, axisIter);
+    if (axis1 != axis2) {
+      [desc transposeDimension:axis1 withDimension:axis2];
+    }
+  }
+  C10_CLANG_DIAGNOSTIC_PUSH()
+#if C10_CLANG_HAS_WARNING("-Wnonnull")
+  C10_CLANG_DIAGNOSTIC_IGNORE("-Wnonnull")
```
Comment (albanD) [inline aten/src/ATen/native/mps/OperationUtils.mm:414]: Why is this necessary?

File: test/test_mps.py (hunk 1 above)
Comment (DenisVieriu97) [reply test/test_mps.py:3573]: Thanks - this is some leftover debug code. Will update

File: aten/src/ATen/native/mps/operations/Indexing.mm (hunk 3 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Indexing.mm:611]: I will update with the issue number

File: aten/src/ATen/native/mps/operations/Copy.mm (hunk 4 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/operations/Copy.mm:92]: Copying from a MPS buffer to a CPU buffer should be done using a single blit copy, without making an extra contiguous buffer, copy into that, then scatter back in the original CPU buffer

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 5 above)
Comment (DenisVieriu97) [reply aten/src/ATen/native/mps/OperationUtils.mm:414]: MPSNDArray **arrayViewWithCommandBuffer:descriptor:aliasing** expects a valid Command Buffer to be passed in, the function didn't not mark the command buffer as an optional argument.
The reason for passing a command buffer is to allow MPS to reshape/materialize the memory if needed (thus encoding work to submit to the GPU on that command buffer). For view ops we shouldn't ever materialize the memory (as long as new new format can be represented using strides), so if MPS would try to make a copy for a view op, we would rather assert out so we catch the issue and fix it.
Note that I've run all torchbench networks / test_mps and test_modules and I haven't seen this issue.

File: test/test_mps.py (hunk 2 above)
Comment (skotapati) [reply test/test_mps.py:3550]: These tests are needed in addition to the opInfo tests, I'll parametrize them as you reccomended

Comment (skotapati) [reply test/test_mps.py]: I've updated the tests, but I can't get parametrize to work. It keeps returning a missing argument error

**Review by albanD — APPROVED** [review]
Ok!
